### DISCUSSION TITLE
Uno single project doesen't work correctly on Rider

### DISCUSSION THREAD

**Discussion by Takoooooo** [unoplatform/uno#16635] [General]
Unfortunately, I couldn't get to import single project Uno project into the existing solution. It compiles, but fails to run and missing ios and android heads in Rider. Is there tutorial how to achieve that? Is it possible?

Created from https://x.com/DinkoPavicic/status/1788277174673277081 in order to help dealing with that problem.

**Comment by Takoooooo** [comment]
@dpavicic If you couldn't get single project to import to solution than try to create it from CLI by `dotnet new unoapp` and then importing it to the solution, that works on my machine, creating it from Rider is blocked by [this issue](https://youtrack.jetbrains.com/issue/RIDER-112161/NRE-when-creating-project-from-template-with-Rider-EAP-2024.1.1).
If you are willing to run the single project on Android you will need to follow [this](https://www.jetbrains.com/help/rider/Run_Debug_Configuration_Xamarin_Android.html) doc. IOS should work out of box but if you have any problems with it, please share more details

**Reply by jeromelaban** [reply]
We're also tracking other rider issues in https://github.com/unoplatform/uno/issues/15226.

**Comment by dpavicic** [comment]
Hi!,
First, I want to thank you for your quick reply. 

I will try to explain in detail what I want to achieve and what's happening...

## What I Want to Accomplish

I have an existing solution with applications and libraries. There are several backend applications (API, Orleans server, Background Worker) and a frontend Blazor Wasm application. I start/debug them simultaneously using Rider's Multi-Launch feature (I would also incorporate Aspire later on).

This works well; I can start and debug them all simultaneously, end-to-end (front end and back end). This is a great experience. Therefore, I would like to add the Uno project to this solution so I can have the same dev experience while developing native apps with Uno.

Unfortunately, I didn't find any way so far to achieve this.

In this post, I will describe the whole process of importing the Uno project into my solution and explain what happens.


## What Would Be Expected

To easily import the Uno project into the existing solution.

  

This is basic information about env:

*   Uno Version: 5.2.108
*   .NET Version: 8
*   Rider Version: Build #RD-241.14494.307, built on April 8, 2024

  

[The repository with all that I described is here](https://github.com/dpavicic/Research.Uno) (solution with imported Uno project and all below steps applied)
  

These are the steps:


## Creating Uno Solution

I created the Uno solution in a separate directory with this command:

```
dotnet new unoapp -o Simple.Uno.App -preset "recommended" -markup "csharp" -theme-service  -theme "fluent" -presentation "mvvm" -config  -di  -log "none" -nav "regions" -http  -loc  -tests "none" -toolkit  -dsp False -id "com.dinkopavicic.Simple.Uno.App" -pub "O=Simple.Uno.App"
```
 

### Opening Uno Solution Created with Command Line

Right from the start, there were problems like this: All .NET Launch Settings Profiles have the **.net8.0-ios Target framework** set. I had to set the correct frameworks for each profile manually.

![image](https://github.com/unoplatform/uno/assets/4416293/ef894070-ecb2-45cd-8f79-f1892a54c85c)
  

### First Build

On the first build, Uno errors with:

`uno_splash_image.xml(2): Error APT2260 : resource drawable/splash_screen (aka com.dinkopavicic.GameSimulator.UnoApp:drawable/splash_screen) not found.`

  

### Second Build

On a second build, the build passes with these Android warnings:

```pgsql
AndroidManifest.xml: Warning XA4218 : Unable to find //manifest/application/uses-library at path: C:\Program Files (x86)\Android\android-sdk\platforms\android-34\optional\androidx.window.extensions.jar
0>AndroidManifest.xml: Warning XA4218 : Unable to find //manifest/application/uses-library at path: C:\Program Files (x86)\Android\android-sdk\platforms\android-34\optional\androidx.window.sidecar.jar
```
  

From this point, I can build and run Uno Desktop, Blazor WASM, and Android (I'm on a Windows machine).
  

**NOTE: Errors and issues I described so far occur on a regular basis when creating a new Uno solution with the command line. In my case, they are 100% reproducible.**

  
### Transferring Uno Project to Existing Solution

Okay, now I want to copy the Uno project into my existing solution. In this solution, I have backend projects and a Blazor Wasm app, and I want the Uno project to be there for native apps so I can start and debug them without a hassle (end-to-end). I want to be able to debug frontends and backends simultaneously. This works really well with Blazor (Wasm, in my case) and [ASP.NET](http://ASP.NET) backend apps.

So, with that in mind, and from what I see how Uno solution is set, I then:

*   Copy the Uno project into my existing solution
*   Update `Directory.Build.props` by adding:

```plain
<!--
      Adding NoWarn to remove build warnings
      NU1507: Warning when there are multiple package sources when using CPM with no source mapping
      NETSDK1201: Warning that specifying RID won't create self containing app
      PRI257: Ignore default language (en) not being one of the included resources (eg en-us, en-uk)
    -->
    <NoWarn>$(NoWarn);NU1507;NETSDK1201;PRI257</NoWarn>
```

and

```plain
<!-- See https://aka.platform.uno/using-uno-sdk#implicit-packages for more information regarding the Implicit Packages version properties. -->
  <PropertyGroup>
    <UnoExtensionsVersion>4.1.14</UnoExtensionsVersion>
    <UnoToolkitVersion>6.0.18</UnoToolkitVersion>
    <UnoThemesVersion>5.0.13</UnoThemesVersion>
    <UnoCSharpMarkupVersion>5.2.13</UnoCSharpMarkupVersion>
  </PropertyGroup>
```

The rest is the same as in the `Directory.Build.props` of my existing solution.

*   Similarly, update the `Directory.Build.targets` and `Directory.Packages.props`
*   Update global.json. I added the `msbuild-sdks` part like this:

```plain
{
  "sdk": {
    "version": "8.0.100",
    "rollForward": "latestMinor",
    "allowPrerelease": false
  },
  // To update the version of Uno please update the version of the Uno.Sdk here. See https://aka.platform.uno/upgrade-uno-packages for more information.
  "msbuild-sdks": {
    "Uno.Sdk": "5.2.108"
  }
}
```

*   Then I update the nuget.config (nothing special here)
*   Then, I opened my existing solution and imported the Uno project into it. Project imports without errors while importing.
*   Now, I need to update the build configurations in the .sln file of the existing solution. I have to add these:

```plain
GlobalSection(SolutionConfigurationPlatforms) = preSolution
		Debug|Any CPU = Debug|Any CPU
		Debug|ARM = Debug|ARM
		Debug|ARM64 = Debug|ARM64
		Debug|iPhone = Debug|iPhone
		Debug|iPhoneSimulator = Debug|iPhoneSimulator
		Debug|x64 = Debug|x64
		Debug|x86 = Debug|x86
		Release|Any CPU = Release|Any CPU
		Release|ARM = Release|ARM
		Release|ARM64 = Release|ARM64
		Release|iPhone = Release|iPhone
		Release|iPhoneSimulator = Release|iPhoneSimulator
		Release|x64 = Release|x64
		Release|x86 = Release|x86
	EndGlobalSection
	GlobalSection(ProjectConfigurationPlatforms) = postSolution
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|Any CPU.Build.0 = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|Any CPU.Deploy.0 = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|ARM.ActiveCfg = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|ARM.Build.0 = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|ARM.Deploy.0 = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|ARM64.ActiveCfg = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|ARM64.Build.0 = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|ARM64.Deploy.0 = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|iPhone.ActiveCfg = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|iPhone.Build.0 = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|iPhone.Deploy.0 = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|iPhoneSimulator.ActiveCfg = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|iPhoneSimulator.Build.0 = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|iPhoneSimulator.Deploy.0 = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|x64.ActiveCfg = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|x64.Build.0 = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|x64.Deploy.0 = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|x86.ActiveCfg = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|x86.Build.0 = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Debug|x86.Deploy.0 = Debug|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|Any CPU.Build.0 = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|Any CPU.Deploy.0 = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|ARM.ActiveCfg = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|ARM.Build.0 = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|ARM.Deploy.0 = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|ARM64.ActiveCfg = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|ARM64.Build.0 = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|ARM64.Deploy.0 = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|iPhone.ActiveCfg = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|iPhone.Build.0 = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|iPhone.Deploy.0 = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|iPhoneSimulator.ActiveCfg = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|iPhoneSimulator.Build.0 = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|iPhoneSimulator.Deploy.0 = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|x64.ActiveCfg = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|x64.Build.0 = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|x64.Deploy.0 = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|x86.ActiveCfg = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|x86.Build.0 = Release|Any CPU
		{61973F6D-07E6-433E-961A-78B19A79993A}.Release|x86.Deploy.0 = Release|Any CPU
	EndGlobalSection
```

and update all GUIDs to be the same as the imported Uno project.

  
Then I restart the solution, but:

*   There are no iOS, Android, and Mac heads in the drop-down menu to choose to build (Rider Build #RD-241.14494.307, built on April 8, 2024)
*   There are only Desktop, Wasm, and WinAppSdk Unpacked, and neither of them runs. They build but fail at runtime with errors like this:

```plain
A fatal error was encountered. The library 'hostpolicy.dll' required to execute the application was not found in 'D:\Work\research\Research-UnoPlatform\src\Simple.Uno.App\bin\Debug\net8.0\'.
Failed to run as a self-contained app.
  - The application was run as a self-contained app because 'D:\Work\research\Research-UnoPlatform\src\Simple.Uno.App\bin\Debug\net8.0\Simple.Uno.App.runtimeconfig.json' was not found.
  - If this should be a framework-dependent app, add the 'D:\Work\research\Research-UnoPlatform\src\Simple.Uno.App\bin\Debug\net8.0\Simple.Uno.App.runtimeconfig.json' file and specify the appropriate framework.
```
  

This is what I have in run/debug configurations:
![image](https://github.com/unoplatform/uno/assets/4416293/b0a90b11-870b-4b29-9944-f56cacafba9a)

Thanks again for the quick reply! 

Cheers, Dinko

**Reply by jeromelaban** [reply]
Thanks for the very detailed report!

> uno_splash_image.xml(2): Error APT2260 : resource drawable/splash_screen (akaom.dinkopavicic.GameSimulator.UnoApp:drawable/splash_screen) not found.

About this error, this is something we're about to release a fix for, you may and see if it helps by adding the following to your `Directory.Build.props` :
```xml
<PropertyGroup>
   <UnoResizetizerVersion>1.5.0-dev.58</UnoResizetizerVersion>
</PropertyGroup>
```

> There are only Desktop, Wasm, and WinAppSdk Unpacked, and neither of them runs.

You're hitting [this particular rider issue](https://youtrack.jetbrains.com/issue/RIDER-111470), for which we do not have a fix for at this time. Additionally, the launch profiles you are seeing need to be aligned with their respective target frameworks (which may not appear because of the rider missing feature). The error appearing also means that the `net8.0` target framework is being executed, and it's not meant to as it's for unit tests only.  We encourage you to upvote the Rider issue so this gets fixed quicker.

As a workaround, [you can use vscode](https://platform.uno/docs/articles/get-started-vscode.html), which supports those targets properly.
